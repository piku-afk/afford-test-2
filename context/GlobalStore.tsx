import {
  Dispatch,
  FC,
  PropsWithChildren,
  useContext,
  useReducer,
  createContext,
  useEffect,
} from 'react';
import { initialState } from './initialState';
import { ActionInterface, ActionTypes, reducer } from './reducer';

interface GlobalContextType {
  state: typeof initialState;
  dispatch: Dispatch<ActionInterface>;
}

export const GlobalContext = createContext<GlobalContextType>({
  state: initialState,
  dispatch: () => initialState,
});

export const useGlobalStore = () => useContext(GlobalContext);

type GlobalStoreProps = {};

export const GlobalStore: FC<PropsWithChildren<GlobalStoreProps>> = (props) => {
  const { children } = props;
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    const init = async () => {
      const res = await fetch('http://localhost:5000/filters');
      const data = await res.json();
      const { searchFilter } = data || {};
      const { brands, categories } = searchFilter || {};

      dispatch({ type: ActionTypes.setCategories, payload: categories });
      dispatch({ type: ActionTypes.setBrands, payload: brands });
    };
    init();
  }, []);

  const value = { state, dispatch };

  return (
    <GlobalContext.Provider value={value}>{children}</GlobalContext.Provider>
  );
};
