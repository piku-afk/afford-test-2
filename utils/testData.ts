export const mockCategories = [
  {
    ID: '1',
    name: 'Hello',
    subCategories: [
      { ID: '1', name: 'Piyush', subCategories: [] },
      { ID: '2', name: 'Mahato', subCategories: [] },
    ],
  },
  { ID: '2', name: 'World 2', subCategories: [] },
];

export const mockProducts = [
  {
    id: '1',
    title: 'product 1',
    description: 'some description',
    price: '',
    category: '',
    image: 'http://placeimg.com',
    quantity: 1,
    rating: { rate: '', count: '' },
  },
];

export const mockBrands = [
  { ID: '1', name: 'hello' },
  { ID: '2', name: 'world' },
];
