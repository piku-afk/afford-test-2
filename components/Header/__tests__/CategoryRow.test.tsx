import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as GlobalContext2 from '@/context/GlobalStore';
import { CategoryRow } from '../categoryRow';
import { mockCategories } from '@/utils/testData';

const user = userEvent.setup();

describe('CategoryRow', () => {
  beforeEach(() => {
    jest.spyOn(GlobalContext2, 'useGlobalStore').mockImplementation(() => ({
      state: { categories: mockCategories, brands: [], cart: [] },
      dispatch: jest.fn(),
    }));
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<CategoryRow />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders different category buttons', () => {
    render(<CategoryRow />);

    const categoryButtons = screen.getAllByRole('button');
    const helloButton = screen.getByRole('button', { name: /hello/i });
    const worldButton = screen.getByRole('button', { name: /world/i });

    expect(categoryButtons).toHaveLength(2);
    expect(helloButton).toBeInTheDocument();
    expect(worldButton).toBeInTheDocument();
  });

  it('renders a popup showing sub-categories', async () => {
    render(<CategoryRow />);

    const helloButton = screen.getByRole('button', { name: /hello/i });
    await user.click(helloButton);

    const categoryPopup = screen.getByRole('presentation');
    const subCategoryButtons = screen.getAllByRole('menuitem');

    expect(categoryPopup).toBeInTheDocument();
    expect(subCategoryButtons).toHaveLength(2);
  });
});

export {};
