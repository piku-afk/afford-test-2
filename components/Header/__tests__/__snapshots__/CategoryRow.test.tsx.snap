// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryRow matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-5x9whe"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      style="padding-left: 48px; padding-right: 48px;"
    >
      <div
        class="css-w4z10b-MuiStack-root"
      >
        <button
          aria-haspopup="true"
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-disableElevation MuiButtonBase-root css-1u7wl2p-MuiButtonBase-root-MuiButton-root"
          id="Hello-menu-button"
          tabindex="0"
          type="button"
        >
          Hello
          <span
            class="MuiButton-endIcon MuiButton-iconSizeSmall css-155nyw6-MuiButton-endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-haspopup="true"
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-disableElevation MuiButtonBase-root css-1u7wl2p-MuiButtonBase-root-MuiButton-root"
          id="World 2-menu-button"
          tabindex="0"
          type="button"
        >
          World 2
          <span
            class="MuiButton-endIcon MuiButton-iconSizeSmall css-155nyw6-MuiButton-endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
